### PR TITLE
Insert ignores at the top of template tags

### LINKED ISSUE #43: Does incorrect thing in gjs/gts files
in gjs/gts files, `{{! template-lint-disable ... }}`  will be added to the top of the file.
This is not correct behavior as the disable can only be at the top of the template, or assigned to the node.

The content-tag-utils' Transformer may be able to help with modifying template-tag contents: 
https://github.com/NullVoxPopuli/content-tag-utils#transformer

### PR DESCRIPTION
**Pull request by wagenet**
Fixes #43.

**DISCLAIMER: This was written with AI. Review accordingly.**

### CODE REVIEW

File: lib/ignore-error.js [hunk 1]
```diff
@@ -1,8 +1,81 @@
...
+      const transformer = new Transformer(file);
+
+      transformer.map((templateContent) => {
+        // Get errors specific to this template block by matching the error source with the template content
+        const templateErrors = errors.filter(error => {
+          // Check if the error's source code appears in this template block
+          const errorLine = error.source?.trim();
+          if (!errorLine) return false;
+
+          return templateContent.includes(errorLine);
```
Comment (NullVoxPopuli) [inline lib/ignore-error.js:64]: this isn't specific enough, we need to check coordinates here the coordinates are passed as the second argument to the callback passed to `map`:
```js
t.map((contents, coordinates) => {
    /*           ^ here ... */
    return 'new content 2';
});
```
https://github.com/NullVoxPopuli/content-tag-utils/blob/main/src/transformer.js#L192C12-L192C21

Comment (NullVoxPopuli) [reply lib/ignore-error.js:64]: Tho.... if the line is an error in both locations... maybe it doesn't matter. 🤔

File: test/ignore.test.js [hunk 2]
```diff
@@ -66,4 +66,113 @@ describe('ignore function', function () {
...
+
+    expect(project.files.app['template.gjs']).to.equal(`import RouteTemplate from 'ember-route-template';
+
+export default RouteTemplate(<template>
+  {{! template-lint-disable no-log }}
+  {{log "hello"}}
+</template>);`);
+  });
+
+  it('should add ignore declarations at the top of multiple template blocks', async function() {
```
Comment (NullVoxPopuli) [inline test/ignore.test.js:112]: nice!

File: test/ignore.test.js [hunk 3]
```diff
@@ -66,4 +66,113 @@ describe('ignore function', function () {
       'ignore-me.hbs': '{{log "ignored"}}',
     });
   })
+
```
Comment (NullVoxPopuli) [inline test/ignore.test.js:69]: can you add this a test?:


```gjs
export const Y = 
  <template>
    <Logger as |log|>
      {{log "foo"}}
    </Logger>
  </template>;

export const X = 
  <template>
    <div>
      {{log "foo"}}
    </div>
  </template>;
  
```

Should result in:
```gjs
export const Y = 
  <template>
    <Logger as |log|>
      {{log "foo"}}
    </Logger>
  </template>;

export const X = 
  <template>
    {{! template-lint-disable no-log }}
    <div>
      {{log "foo"}}
    </div>
  </template>;

```

File: lib/ignore-error.js [hunk 4]
```diff
@@ -1,33 +1,114 @@
...
+      newLines.push(line);
+    }
+
+    return newLines.join('\n');
   }
+}
 
-  const firstLine = file.split('\n')[0];
+function isErrorInTemplateBlock(error, templateContent, coordinates, fileContent) {
+  if (!error.line || !error.column) return false;
```
Comment (mansona) [inline lib/ignore-error.js:52]: this line looks a bit strange. This is essentially saying that if the error has no line or column information then we always think it's "not in a template block" and it will silently be ignored 🤔 

I feel like this is a bigger error if that's true, do we know any cases where this could be true? should we throw here if it happens?

File: lib/ignore-error.js [hunk 5]
```diff
@@ -1,33 +1,114 @@
...
 
-  if (firstLine.includes('template-lint-disable')) {
-    const matched = firstLine.match(/template-lint-disable(.*)(--)?\}\}/);
-    const existing = matched[1].split(' ')
-      .map(item => item.trim())
-      .filter(item => item.length);
+  // Get the error's character offset by counting characters up to the error position
+  const lines = fileContent.split('\n');
+  const errorOffset = lines.slice(0, error.line - 1).reduce((sum, line) => sum + line.length + 1, 0) + error.column;
```
Comment (mansona) [inline lib/ignore-error.js:56]: this feels like a very complex one liner that could do with a bit more explanation. Why are we using a reduce in here? can you describe the need for this complexity in a comment or something?

File: lib/ignore-error.js [hunk 6]
```diff
@@ -1,33 +1,114 @@
...
+          return addIgnoreToContent(templateContent, mergedRules);
+        } else {
+          return addIgnoreToContent(templateContent, ruleIds);
+        }
+      });
 
-    writeFileSync(filePath, file.replace(/^.*\n/, `{{! template-lint-disable ${uniqueIds.join(' ')} }}\n`));
+      writeFileSync(filePath, transformer.toString());
+    } catch {
+      console.warn("Unable to parse template file", filePath);
```
Comment (mansona) [inline lib/ignore-error.js:95]: shouldn't this be an error?

**Review by mansona — COMMENTED** [review]
This is mostly ok, you should rebase now that we've dropped ember-template-lint@2 support 👍 

I have left a few comments/questions in the review but I'm probably leaning towards merging this since it's fully tested

**Review by gitKrystan — APPROVED** [review]

**Comment by mansona** [comment]
So this is great, I like some of the more recent changes and the tests look good... but my original questions haven't been answered 🤔 and I double checked that they are still relevant with the changes since my review and the code is still there. Can someone take a look?

File: lib/ignore-error.js [hunk 7]
```diff
@@ -1,33 +1,132 @@
...
+        addedIgnore = true;
+      }
+      newLines.push(line);
+    }
+
+    return newLines.join('\n');
+  }
+}
+
+function isErrorInTemplateBlock(error, templateContent, coordinates, fileContent) {
```
Comment (BlueCutOfficial) [inline lib/ignore-error.js]: `templateContent` is unused, we use only the coordinates.

File: lib/ignore-error.js [hunk 8]
```diff
@@ -1,33 +1,132 @@
...
-    uniqueIds.sort((a, b) => a.localeCompare(b));
+  // Convert line/column to character offset to match against template block coordinates
+  // Sum all characters in lines before error (including newlines) plus column offset
+  const lines = fileContent.split('\n');
+  const errorOffset = lines.slice(0, error.line - 1).reduce((sum, line) => sum + line.length + 1, 0) + error.column;
 
-    writeFileSync(filePath, file.replace(/^.*\n/, `{{! template-lint-disable ${uniqueIds.join(' ')} }}\n`));
+  // Check if the error offset is within the template block's range
+  return errorOffset >= coordinates.start && errorOffset <= coordinates.end;
+}
```
Comment (BlueCutOfficial) [inline lib/ignore-error.js]: I think it's still not quite clear what is happening. I would suggest moving the `slice` above and rewriting it this way: 

```js
/*
   * The coordinates are the template block's range in the fileContent
   * = where this specific part of template starts and ends.
   * We want to know if the offset where the error starts is within that range.
   */
  // Get the lines of the fileContent from start to the latest line before the errorred line
  const linesToError = fileContent.split('\n').slice(0, error.line - 1);
  // Summing up all the characters gives us the exact offset where the error starts.
  const errorOffset = linesToError.reduce((sum, line) => sum + line.length + 1, 0) + error.column;
  // Check the result against the template block's range
  return errorOffset >= coordinates.start && errorOffset <= coordinates.end;
```

💡 I think this function is not optimized at the moment. 
Most of the function - from first line to `reduce` - is unrelated to the current template block, it's about the whole file content. So we are recalculating the error offset in the file for each template block it contains. Couldn't these things be done before the transformer mapping?
It's maybe a bit edge case but that scenario is highlighted by the Logger+regular log test.

File: lib/ignore-error.js [hunk 9]
```diff
@@ -1,33 +1,132 @@
...
-  if (firstLine.includes('template-lint-disable')) {
-    const matched = firstLine.match(/template-lint-disable(.*)(--)?\}\}/);
-    const existing = matched[1].split(' ')
-      .map(item => item.trim())
-      .filter(item => item.length);
+function createIgnoreDeclaration(rules, indentation = '') {
+  return `${indentation}{{! template-lint-disable ${rules.join(' ')} }}`;
+}
+
+function addIgnoreToContent(content, rules) {
```
Comment (BlueCutOfficial) [inline lib/ignore-error.js]: I think using the term "content tag" in the function would make it clearer that this function is about the `<template>` case specifically.

```suggestion
function addIgnoreToContentTag(content, rules) {
```

File: lib/ignore-error.js (hunk 5 above)
Comment (mansona) [reply lib/ignore-error.js]: turns out we didn't need this complexity at all 🤷 thinking about it for 5 mins made me see that this is way simpler 🙈

File: lib/ignore-error.js (hunk 6 above)
Comment (mansona) [reply lib/ignore-error.js]: Sorry I didn't know this was fixed in a new commit, it was really hard to follow along with what was going on with the commit history so I've squashed it and it makes more sense 👍

File: lib/ignore-error.js (hunk 4 above)
Comment (mansona) [reply lib/ignore-error.js]: not having a line for where the error is reported just isn't something that we need to guard against here, it's not reasonable to assume this would be the case. I've removed this check and made it only rely on the line number

File: lib/ignore-error.js (hunk 7 above)
Comment (mansona) [reply lib/ignore-error.js]: yes you are right, but it turns out the whole function would be much simpler if we used the templateContent and not the fileContent 😂 I swapped it around and removed the unused stuff 👍

File: lib/ignore-error.js (hunk 8 above)
Comment (mansona) [reply lib/ignore-error.js]: take a look at what I did and see how much simpler it is now, we just don't need to track lines and columns at all 🙈 it was all way overkill

File: lib/ignore-error.js (hunk 9 above)
Comment (mansona) [reply lib/ignore-error.js]: you are right, but this actually inspired me to think "why are we treating them different". It's a bit restructured now but I ended up re-using this same function in both the content-tag and hbs cases. I also renamed it to be a bit clearer 👍

**Review by mansona — APPROVED** [review]
I feel like I should be saying "sorry for the delay" here, but to be brutally honest this PR took so long because it was vibe coded 🤷 

If a human sat down and wrote this with maintainability in mind it would have been merged months ago.
